lib/THC/THCGeneral.h

```diff
@@ -35,37 +35,45 @@
 #endif
 
 struct THCRNGState;  /* Random number generator state. */
-struct THCBlasState;
+
+typedef struct _THCCudaResourcesPerDevice {
+  cudaStream_t* streams;
+  cublasHandle_t* blasHandles;
+  /* Size of scratch space per each stream on this device available */
+  size_t scratchSpacePerStream;
+  /* Device-resident scratch space per stream, used for global memory
+     reduction kernels. */
+  void** devScratchSpacePerStream;
```

Comment:
I added that for the global reduction kernels. I follow the same strategy as Thrust: if the reduction is small, it's done in one block. If the reduction is large, two kernel launches/passes are used: each block does a reduction and writes back to a global memory scratch space, and a second kernel is launched to reduce from that.

Thrust allocates and deallocates the scratch buffer for every call, whereas I keep the scratch buffer space resident. The scratch space is only ~# of SMs * 4 * sizeof(float) bytes per stream, so relatively small.

This global memory scratch space can be used for other purposes too; anywhere where you need to perform two passes of work with not that much data needing to remain resident between the two passes.

It's managed here because we need independent space for each stream.

Also, I should convert logicalall and logicalany to this new reduction stuff as well, since those were added after I added the reduction stuff locally.